Reference documentation for the CSS Custom highlight API

captainbrosset opened the pull request:
<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Description

This PR contains the reference documentation for the `CSS Custom Highlight API`.

### Motivation

The API is now available in Chromium-based browser (Chrome 105+, Edge 105+, etc.), and is behind a flag in Safari.

### Additional details

I wrote this blog post last year on css-tricks about it: https://css-tricks.com/css-custom-highlight-api-early-look/

The spec: https://www.w3.org/TR/css-highlight-api-1/

### Related issues and pull requests

This fixes https://github.com/mdn/mdn/issues/348

It is also related to this BCD PR (which should ideally land before): https://github.com/mdn/browser-compat-data/pull/18720

captainbrosset committed:
Started documenting the custom highlight API

captainbrosset committed:
Merge remote-tracking branch 'origin/main' into custom-highlight

captainbrosset committed:
Completed the highlight and highlightregistry ref docs

dipikabh commented:
Hi @wbamberg, would you be able to review this?

wbamberg commented:
Sure, I'd be happy to :)

dipikabh commented:
Thanks, @wbamberg!

wbamberg commented on files/en-us/web/api/highlightregistry/get/index.md:
```diff
@@ -0,0 +1,57 @@
+
+## Examples
+
+The following code sample demonstrates how to create a new `Highlight`, set it in the registry, and retrieve it by its name using the `get()` method:
```
```suggestion
The following code sample demonstrates how to create a new `Highlight`, add it to the registry, and retrieve it by its name using the `get()` method:
```

wbamberg commented on files/en-us/web/api/highlightregistry/has/index.md:
```diff
@@ -0,0 +1,56 @@
+
+{{APIRef("CSS Custom Highlight API")}}
+
+The **`has()`** method of the {{domxref("HighlightRegistry")}} interface returns a boolean indicating whether a {{domxref("Highlight")}} object with the specified name exists in registry or not.
```
```suggestion
The **`has()`** method of the {{domxref("HighlightRegistry")}} interface returns a boolean indicating whether or not a {{domxref("Highlight")}} object with the specified name exists in the registry.

wbamberg commented on files/en-us/web/api/highlightregistry/has/index.md:
```diff
@@ -0,0 +1,56 @@
+
+Returns `true` if a highlight with the specified name exists in the registry; otherwise `false`.
+
+## Example
```
```suggestion
## Examples
```

wbamberg commented on files/en-us/web/api/highlightregistry/keys/index.md:
```diff
@@ -0,0 +1,78 @@
+
+{{APIRef("CSS Custom Highlight API")}}
+
+The **`keys()`** method of the {{domxref("HighlightRegistry")}} interface returns a new [Iterator](/en-US/docs/Web/JavaScript/Guide/Iterators_and_Generators) object that
+contains the keys for each `Highlight` object in the `HighlightRegistry` object in insertion order.
```
```suggestion
The **`keys()`** method of the {{domxref("HighlightRegistry")}} interface returns a new [Iterator](/en-US/docs/Web/JavaScript/Guide/Iterators_and_Generators) object that contains the names of each `Highlight` object in the `HighlightRegistry` object in insertion order.
```

wbamberg commented on files/en-us/web/api/highlightregistry/keys/index.md:
```diff
@@ -0,0 +1,78 @@
+
+### Return value
+
+A new iterator object containing the keys for each `Highlight` object in the registry, in insertion order.
```
```suggestion
A new iterator object containing the names of each `Highlight` object in the registry, in insertion order.
```

wbamberg commented on files/en-us/web/api/highlightregistry/set/index.md:
```diff
@@ -0,0 +1,71 @@
+### Parameters
+
+- `name`
+  - : The name of the `Highlight` object to add to the registry. The name must be a {{jsxref("String")}}.
```
```suggestion
  - : The name of the `Highlight` object to add or update. The name must be a {{jsxref("String")}}.
```

wbamberg commented on files/en-us/web/api/highlightregistry/set/index.md:
```diff
@@ -0,0 +1,71 @@
+- `name`
+  - : The name of the `Highlight` object to add to the registry. The name must be a {{jsxref("String")}}.
+- `highlight`
+  - : The `Highlight` object to add to the registry. This must be a {{domxref("Highlight")}} interface instance.
```
```suggestion
  - : The `Highlight` object to add or update. This must be a {{domxref("Highlight")}} interface instance.
```

wbamberg commented on files/en-us/web/api/highlightregistry/set/index.md:
```diff
@@ -0,0 +1,71 @@
+
+### Return value
+
+The `HighlightRegistry` object is returned.
```
```suggestion
The `HighlightRegistry` object.
```

wbamberg commented on files/en-us/web/api/highlightregistry/size/index.md:
```diff
@@ -0,0 +1,47 @@
+
+{{APIRef("CSS Custom Highlight API")}}
+
+The **`size`** property returns the number of {{domxref("Highlight")}} objects in a {{domxref("HighlightRegistry")}}.
```
```suggestion
The **`size`** property returns the number of {{domxref("Highlight")}} objects in the {{domxref("HighlightRegistry")}}.
```

(we use "the" in the other descriptions)

wbamberg commented on files/en-us/web/api/highlightregistry/size/index.md:
```diff
@@ -0,0 +1,47 @@
+
+## Value
+
+The value of `size` is a read-only integer representing how many entries the {{jsxref("Map")}}-like `HighlightRegistry` object has.
```
```suggestion
A read-only integer indicating how many `Highlight` objects the registry contains.
```

wbamberg commented on files/en-us/web/api/highlightregistry/values/index.md:
```diff
@@ -0,0 +1,78 @@
+
+{{APIRef("CSS Custom Highlight API")}}
+
+The **`values()`** method of the {{domxref("HighlightRegistry")}} interface returns a new [Iterator](/en-US/docs/Web/JavaScript/Guide/Iterators_and_Generators) object that
+contains the values for each `Highlight` object in the `HighlightRegistry` object in insertion order.
```
```suggestion
The **`values()`** method of the {{domxref("HighlightRegistry")}} interface returns a new [Iterator](/en-US/docs/Web/JavaScript/Guide/Iterators_and_Generators) object that contains each `Highlight` object in the registry, in insertion order.
```

wbamberg commented on files/en-us/web/api/highlightregistry/values/index.md:
```diff
@@ -0,0 +1,78 @@
+
+### Return value
+
+A new iterator object containing the values for each `Highlight` object in the registry, in insertion order.
```
```suggestion
A new iterator object containing each `Highlight` object in the registry, in insertion order.
```

(but is "contains" right? Should we say "yields"? here and also in the similar `Highlight` methods?)

wbamberg commented on files/en-us/web/css/_doublecolon_highlight/index.md:
```diff
@@ -0,0 +1,74 @@
+
+## Examples
+
+### HTML
```
Would love to see a live sample here.

wbamberg commented on files/en-us/web/css/_doublecolon_highlight/index.md:
````diff
@@ -0,0 +1,74 @@
+<p id="foo">This text is highlighted with the CSS Custom Highlight API.</p>
+```
+
+### JavaScript
````
Super-picky, but the order of code blocks should be: HTML, CSS, JavaScript (https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Page_structures/Code_examples#formatting_live_samples). And yes, humans should not have to take care of this :(.

wbamberg commented on files/jsondata/GroupData.json:
```diff
@@ -163,6 +163,13 @@
+      "overview": ["CSS Custom Highlight API"],
+      "interfaces": ["Highlight", "HighlightRegistry"],
+      "methods": [],
+      "properties": [],
```
Should this include CSS.highlights?

captainbrosset replied:
Yes, great point. In fact I should also link most (all?) instances of `CSS.highlights` that I wrote in the new docs to the actual CSS.highlights page.

wbamberg reviewed (changes requested):
Thanks Patrick! I hope the comments are helpful :)

captainbrosset committed:
Merge remote-tracking branch 'origin/main' into custom-highlight

captainbrosset commented on files/en-us/web/api/css_custom_highlight_api/index.md:
```diff
@@ -0,0 +1,242 @@
+tags:
+  - API
+  - CSS Custom Highlight
+  - Reference
+browser-compat:
```
Hmm, actually I have 3 now 😁

wbamberg replied:
ha ha. This should work if you have a recent enough version of Yari, it's dependent on https://github.com/mdn/yari/pull/8023 which got merged 3 days ago. Maybe if we update this branch from main it will pick up the new Yari and then work?

wbamberg replied:
Seems to be good now: https://pr23907.content.dev.mdn.mozit.cloud/en-US/docs/Web/API/CSS_Custom_Highlight_API#specifications :)

captainbrosset replied:
Awesome. I updated the dependencies locally and now it works for me too.

captainbrosset commented on files/en-us/web/api/highlight/add/index.md:
```diff
@@ -0,0 +1,61 @@
+  - CSS Custom Highlight API
+  - Reference
+  - add
+browser-compat: api.Highlight.add
```
The problem is that there is no spec url for `add` and many other methods. The spec says that Highlight is a Set-like object and just doesn't list the methods that come with it (add, clear, delete, forEeach, ...).

Elchi3 replied:
hm, I'm not sure what our policy is for Map-like and Set-like objects. When I documented WebXR, I didn't create subpages because it seemed repetitive to do that. For example, https://developer.mozilla.org/en-US/docs/Web/API/XRHand mentions the Map-like methods on the main page but doesn't re-explain the individual methods on separate pages. I'm referring to the JS Map docs for that.

captainbrosset replied:
Yeah, I can see both things being viable here. The nice thing about actually creating the sub-pages is that you can provide custom examples that will be more telling than the generic map or set examples. I think the language you use on these sub-pages can also be more specific to the API you're documenting, and that makes it easier to understand.
But, yeah, much more repetitive work. If only we had a thing to generate stubs for those sub-pages (see [this related issue](https://github.com/dontcallmedom/mdn-scaffold/issues/4)).

wbamberg replied:
> The problem is that there is no spec url for `add` and many other methods. The spec says that Highlight is a Set-like object and just doesn't list the methods that come with it (add, clear, delete, forEeach, ...).

Oh, I see. That's unfortunate but understandable!

> The nice thing about actually creating the sub-pages is that you can provide custom examples that will be more telling than the generic map or set examples. I think the language you use on these sub-pages can also be more specific to the API you're documenting, and that makes it easier to understand.

Yes, absolutely. A lot of my comments in this PR are about tweaking the language in this way. It's this classic conundrum where duplicating effort clearly makes things better for developers, but makes more maintenance work.

For the "Specifications" thing, if we do want to keep these duplicated pages, maybe we could just use text like:

```
## Specifications

As for [`Set.prototype.entries()`](https://tc39.es/ecma262/multipage/keyed-collections.html#sec-set.prototype.entries).
```

?

wbamberg replied:
Or we could use `spec-urls`, though it feels a bit uncomfortable to use that in an essentially ad hoc way. But I suppose it's no different from ad hoc text, and means we could validate the URLs.

captainbrosset replied:
I went with this approach. Thanks Will.

Added the map spec-urls to all the map-like pages of the HighlightRegistry interface, and the set spec-urls to all the set-like pages of the Highlight interface.

captainbrosset commented on files/en-us/web/api/highlight/entries/index.md:
````diff
@@ -0,0 +1,68 @@
+console.log(iter.next().value); // [Range, Range]
+```
+
+The following code example shows how to iterate over the ranges in a highlight by using a [`for...of`](/en-US/docs/Web/JavaScript/Reference/Statements/for...of) loop:
+
+```js
+const myHighlight = new Highlight();
+myHighlight.add(new Range());
+myHighlight.add(new Range());
+
+for (const [range] of myHighlight.entries()) {
+  console.log(range); // Range
+}
+```
````
Yes, I think this is a great idea. And TIL: you can create a `Text` object 🤯

captainbrosset commented on files/en-us/web/api/highlight/keys/index.md:
```diff
@@ -0,0 +1,41 @@
+  - CSS Custom Highlight API
+  - Reference
+  - keys
+browser-compat: api.Highlight.keys
```
Yeah, same for all of the set-like methods and properties of `Highlight`, as well as all the map-like methods and properties of `HighlightRegistry`.

captainbrosset replied:
Added the map spec-urls to all the map-like pages of the HighlightRegistry interface, and the set spec-urls to all the set-like pages of the Highlight interface.

captainbrosset commented on files/en-us/web/api/highlightregistry/index.md:
```diff
@@ -0,0 +1,91 @@
+- {{domxref("HighlightRegistry.forEach()")}}
+  - : Calls `callbackFn` once for each range present in the highlight object, in insertion order.
+- {{domxref("HighlightRegistry.get()")}}
+  - : sdfh
```
🤦🏽 Sorry Will, I realize I did no review my own changes in this file. Nice of you to create comments for each one of these lines. You should definitely feel free to tell me to rework the entire section next time 😃

captainbrosset commented on files/en-us/web/api/highlightregistry/clear/index.md:
```diff
@@ -0,0 +1,58 @@
+  - CSS Custom Highlight API
+  - Reference
+  - clear
+browser-compat: api.HighlightRegistry.clear
```
See https://github.com/mdn/content/pull/23907#discussion_r1095568771

captainbrosset replied:
Added the map spec-urls to all the map-like pages of the HighlightRegistry interface, and the set spec-urls to all the set-like pages of the Highlight interface.

captainbrosset commented on files/en-us/web/api/highlightregistry/delete/index.md:
```diff
@@ -0,0 +1,59 @@
+  - CSS Custom Highlight API
+  - Reference
+  - delete
+browser-compat: api.HighlightRegistry.delete
```
See https://github.com/mdn/content/pull/23907#discussion_r1095568771

captainbrosset replied:
Added the set and map spec-urls to all the set-like and map-like pages in Highlight and HighlightRegistry.

captainbrosset commented on files/en-us/web/api/highlight/priority/index.md:
```diff
@@ -0,0 +1,63 @@
+
+When overlapping ranges are used by multiple different {{domxref("Highlight")}} objects, and styled using the {{cssxref("::highlight")}} pseudo-element, the browser needs to decide the order in which to apply these styles.
+
+The `priority` property of the {{domxref("Highlight")}} interface is a {{jsxref("Number")}} used to determine which highlight should be used to style multiple highlights' overlapping parts.
+
+By default, highlights have a priority of `0`, and the browser chooses the most recently registered highlight.
```
This is an awesome example!

I'm rephrasing the intro, adding a simple example to show the default behavior, and then adding your example.

captainbrosset committed:
Addressed review comments

captainbrosset commented:
@wbamberg thanks for the thorough review. I believe I've addressed most comments. The missing BCD tables problem is still there, see my comments. And your suggestion to add a spec-url led to more spec URLs on the page, not fewer :)

wbamberg commented:
Thanks for the updates, @captainbrosset ! If we can resolve the [spec URLs issue](https://github.com/mdn/content/pull/23907#discussion_r1095026206) then I think this is good.

For spec URLs, I think we should use the `spec-urls` front matter key unless @Elchi3 has an alternative.

Elchi3 commented:
> For that issue, I think we should use the `spec-urls` front matter key unless @Elchi3 has an alternative.

I think that's fine for the moment. I just also found https://github.com/mdn/browser-compat-data/issues/6367 in which BCD also needs to figure out what to do about set-like and map-like objects etc. Hopefully we can standardize this one day and make things consistent.

captainbrosset committed:
Merge remote-tracking branch 'origin/main' into custom-highlight

captainbrosset committed:
Added map and set spec-urls to Highlight and HighlightRegistry

captainbrosset commented:
@wbamberg @Elchi3  thanks for your help on this. I added the necessary `spec-urls` to all of the set-like and map-like pages in this PR so they point to the set and map specs.
This should be ready for a final review.

wbamberg reviewed (approved):
👍 looks great, thank you @captainbrosset !

captainbrosset commented:
Thank you for the great reviews and conversations!
